### PR TITLE
Core: Fix epoch millis java time formatter

### PR DESCRIPTION
**Pull request by spinscale**
The existing implemention could not deal with negative numbers as well
as +- 999 milliseconds around the epoch.

This commit uses Instant.ofEpochMilli() and tries to parse the input as
a number directly instead of using a date formatter.

### CODE REVIEW

**Review by cbuescher — CHANGES_REQUESTED** [review]
@spinscale looks good, but I played around with the output of the formatter a bit for negative numbers and ran into some unexpected results. I left an example, maybe you can comment?

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java [hunk 1]
```diff
@@ -879,11 +881,42 @@
...
     /*
      * Returns a formatter for parsing the milliseconds since the epoch
+     * This one needs a custom implementation, because the standard date formatter can not parse negative values
+     * or anything +- 999 milliseconds around the eopch
+     *
+     * This implementation just resorts to parsing the input directly to an Instant by trying to parse a number.
      */
-    private static final CompoundDateTimeFormatter EPOCH_MILLIS = new CompoundDateTimeFormatter(new DateTimeFormatterBuilder()
+    private static final DateTimeFormatter EPOCH_MILLIS_FORMATTER = new DateTimeFormatterBuilder()
         .appendValue(ChronoField.INSTANT_SECONDS, 1, 19, SignStyle.NEVER)
```
[the remark this answers is not in the document]
Comment (spinscale) [reply server/src/main/java/org/elasticsearch/common/time/DateFormatters.java]: Good call. It seems that the date time formatter cannot deal with negative dates... overwriting the `format()` is something we can do in the custom implementation though.

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java [hunk 2]
```diff
@@ -879,11 +881,42 @@
 
     /*
      * Returns a formatter for parsing the milliseconds since the epoch
+     * This one needs a custom implementation, because the standard date formatter can not parse negative values
+     * or anything +- 999 milliseconds around the eopch
```
[the remark this answers is not in the document]
Comment (spinscale) [reply server/src/main/java/org/elasticsearch/common/time/DateFormatters.java]: fixed

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java [hunk 3]
```diff
@@ -879,11 +881,42 @@
...
-    private static final CompoundDateTimeFormatter EPOCH_MILLIS = new CompoundDateTimeFormatter(new DateTimeFormatterBuilder()
+    private static final DateTimeFormatter EPOCH_MILLIS_FORMATTER = new DateTimeFormatterBuilder()
         .appendValue(ChronoField.INSTANT_SECONDS, 1, 19, SignStyle.NEVER)
         .appendValue(ChronoField.MILLI_OF_SECOND, 3)
-        .toFormatter(Locale.ROOT));
+        .toFormatter(Locale.ROOT);
+
+    private static final class EpochDateTimeFormatter extends CompoundDateTimeFormatter {
+
+        EpochDateTimeFormatter() {
```
[the remark this answers is not in the document]
Comment (spinscale) [reply server/src/main/java/org/elasticsearch/common/time/DateFormatters.java]: the class was already private... but fixed it for both ctors two as well

**Review by cbuescher — APPROVED** [review]
Thanks for the change and additional test. LGTM

**Pull request merged**